### ISSUE TITLE
Performance can be increased by removing a number of unnecessary memset calls.

### ISSUE THREAD

**Issue by LeeTibbert** [scala-native/scala-native#2277]
Why zero memory which is already zero?

Due the twists & turns of development, there appear to be a handful (20?) or so places in the SN
code base where a memset() to zero is done immediately after a  unsafe.package alloc, which
is (circa SN 4.0 or just earlier) documented to return cleared memory. 

I noticed this when working on socket structures.  There is no such thing as networking code
which is too fast, so I immediately twinged to waste & redundancy.

It is unfortunate that the name of unsafe.package alloc was not changed at the time, but that
is the subject of both an existing PR and a hoped for Issue containing my thoughts for improvement
there.

By my understanding, the "usual' use of alloc(), meaning unsafe.package alloc, clears memory.
A direct call to Zone.alloc (say be explicitly creating a zone "Phantom", and then explicitly calling
Phantom.alloc) will return memory whose contents are undefined.

I believe a first step in sorting out some of the misdirection, is to remove obviously redundant calls
to memset. In at least one place libc.calloc() can be called instead of a libc.malloc/libc.memset sequence.

**Comment by LeeTibbert** [comment]
In associated PR #2278 there is some room for bike shedding about the movement
of the allocation of the hints structure into the Zone in `java.net.PlainSocketImpl.scala#bind`.

Without squandering performance, one must set the context that tiny differences in memory
allocation are swamped by the trip to network land in the called C bind() routine.

That said, moving the hints provides better locality of reference.  Given the current 
implementation the implied malloc/memset is probably a few CPU cycles 
and a memory allocation slower that the previous stackalloc[]/memset.  
I hope that someday unsafe.alloc has a simple & fast bump allocator for
cleared memory.  Until then, save where you can; spend where you want.

**Comment by LeeTibbert** [comment]
I took a look at re-working the two places in PlainSocketImpl.scala and promptly
ran into number of defects which cause much greater concern than a few
extra memory allocation cycles. I have created Issues for two of these.

My short run recommendation is to go with the code in the PR and spend
cycles studying and fixing the more concerning issues.

The medium run fix is to eliminate some code in each of `bind()` & `connect()`
which is time consuming and unnecessary.  Without having completed a design
sketch, I believe a memory allocation will still be necessary.  I prefer `alloc` but,
depending upon the amount of information needing to be filled in, may
be able to use `stackalloc` (get rid of the overhead of an entire Zone establishment).
